File: src/main/java/org/mitallast/queue/crdt/commutative/GCounter.java
package org.mitallast.queue.crdt.commutative;

import com.google.common.base.Preconditions;
import gnu.trove.impl.sync.TSynchronizedLongLongMap;
import gnu.trove.map.TLongLongMap;
import gnu.trove.map.hash.TLongLongHashMap;
import gnu.trove.procedure.TLongProcedure;
import org.mitallast.queue.common.codec.Codec;
import org.mitallast.queue.common.codec.Message;
import org.mitallast.queue.crdt.replication.Replicator;

/**
 * Using counter vector allows to implement garbage collection
 */
public class GCounter implements CmRDT {

    public static class SourceAssign implements SourceUpdate {
        public static final Codec<SourceAssign> codec = Codec.of(
            SourceAssign::new,
            SourceAssign::value,
            Codec.longCodec
        );

        private final long value;

        public SourceAssign(long value) {
            this.value = value;
        }

        public long value() {
            return value;
        }
    }

    public static class DownstreamAssign implements DownstreamUpdate {
        public static final Codec<DownstreamAssign> codec = Codec.of(
            DownstreamAssign::new,
            DownstreamAssign::replica,
            DownstreamAssign::value,
            Codec.longCodec,
            Codec.longCodec
        );

        private final long replica;
        private final long value;

        public DownstreamAssign(long replica, long value) {
            this.replica = replica;
            this.value = value;
        }

        public long replica() {
            return replica;
        }

        public long value() {
            return value;
        }
    }

    private final long id;
    private final long replica;
    private final Replicator replicator;
    private final TLongLongMap counterMap;

    public GCounter(long id, long replica, Replicator replicator) {
        this.id = id;
        this.replica = replica;
        this.replicator = replicator;
        this.counterMap = new TSynchronizedLongLongMap(new TLongLongHashMap());
    }

    @Override
    public void update(Message event) {
        if (event instanceof SourceUpdate) {
            sourceUpdate((SourceUpdate) event);
        } else if (event instanceof DownstreamUpdate) {
            downstreamUpdate((DownstreamUpdate) event);
        }
    }

    @Override
    public boolean shouldCompact(Message event) {
        return event instanceof DownstreamAssign &&
            ((DownstreamAssign) event).value < counterMap.get(replica);
    }

    @Override
    public void sourceUpdate(SourceUpdate update) {
        if (update instanceof SourceAssign) {
            add(((SourceAssign) update).value);
        }
    }

    @Override
    public synchronized void downstreamUpdate(DownstreamUpdate update) {
        if (update instanceof DownstreamAssign) {
            DownstreamAssign downstream = (DownstreamAssign) update;
            long current = counterMap.get(downstream.replica);
            if (current < downstream.value) {
                counterMap.put(downstream.replica, downstream.value);
            }
        }
    }

    public long increment() {
        return add(1);
    }

    public long add(long value) {
        Preconditions.checkArgument(value >= 0, "must be positive");
        long updated = counterMap.adjustOrPutValue(replica, value, value);
        replicator.append(id, new DownstreamAssign(replica, updated));
        return updated;
    }

    public long value() {
        SumProcedure sum = new SumProcedure();
        counterMap.forEachValue(sum);
        return sum.value;
    }

    private static class SumProcedure implements TLongProcedure {
        long value = 0;

        @Override
        public boolean execute(long value) {
            this.value += value;
            return true;
        }
    }
}

File: src/main/java/org/mitallast/queue/crdt/commutative/GCounter.kt
package org.mitallast.queue.crdt.commutative

import com.google.common.base.Preconditions
import gnu.trove.impl.sync.TSynchronizedLongLongMap
import gnu.trove.map.hash.TLongLongHashMap
import gnu.trove.procedure.TLongProcedure
import org.mitallast.queue.common.codec.Codec
import org.mitallast.queue.common.codec.Message
import org.mitallast.queue.crdt.replication.Replicator

/**
 * Using counter vector allows to implement garbage collection
 */
class GCounter(private val id: Long, private val replica: Long, private val replicator: Replicator) : CmRDT {

    data class SourceAssign(val value: Long) : CmRDT.SourceUpdate {
        companion object {
            val codec = Codec.of(
                ::SourceAssign,
                SourceAssign::value,
                Codec.longCodec()
            )
        }
    }

    data class DownstreamAssign(val replica: Long, val value: Long) : CmRDT.DownstreamUpdate {
        companion object {
            val codec = Codec.of(
                ::DownstreamAssign,
                DownstreamAssign::replica,
                DownstreamAssign::value,
                Codec.longCodec(),
                Codec.longCodec()
            )
        }
    }

    private val counterMap = TSynchronizedLongLongMap(TLongLongHashMap())

    override fun update(event: Message) {
        when (event) {
            is CmRDT.SourceUpdate -> sourceUpdate(event)
            is CmRDT.DownstreamUpdate -> downstreamUpdate(event)
        }
    }

    override fun shouldCompact(event: Message): Boolean {
        return event is DownstreamAssign && event.value < counterMap.get(replica)
    }

    override fun sourceUpdate(update: CmRDT.SourceUpdate) {
        when (update) {
            is SourceAssign -> add(update.value)
        }
    }

    @Synchronized override fun downstreamUpdate(update: CmRDT.DownstreamUpdate) {
        when (update) {
            is DownstreamAssign -> {
                val current = counterMap.get(update.replica)
                if (current < update.value) {
                    counterMap.put(update.replica, update.value)
                }
            }
        }
    }

    fun increment(): Long = add(1)

    fun add(value: Long): Long {
        Preconditions.checkArgument(value >= 0, "must be positive")
        val updated = counterMap.adjustOrPutValue(replica, value, value)
        replicator.append(id, DownstreamAssign(replica, updated))
        return updated
    }

    fun value(): Long {
        val sum = SumProcedure()
        counterMap.forEachValue(sum)
        return sum.value
    }

    private class SumProcedure : TLongProcedure {
        internal var value: Long = 0

        override fun execute(value: Long): Boolean {
            this.value += value
            return true
        }
    }
}
